### PR TITLE
CORE: context_config_read/modify/print/release

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Implementation of CORE UCC functionality: ucc_context_config_read/release
Additionally added ucc_context_config_print and modify to UCC API.

## Why ?
context_config_print is needed both for consistency of the API and for usability (we will use it at least in ucc_info).
context_config_modify is a key functionality when user wants to tweak ctx from the code (not from the env var, use case - multiple rails hca setting). 

## How ?
Note ucc_context is a combination of cl_contexts, hence ucc_context_config is a combination of the corresponding cl_context_configs. Different CL contexts will have different configuration variables. This is why some modify parameters (name, value) will only be applicable to the specific CLS. Because of that the API ucc_context_config_modify also takes a const char* string  - comma separated list of CLs whose cl_contexts will be modified. 

For example:
1. ucc_context_config_modify(cfg, "all", "TEST_PARAM", "aaa") will try to modify the cl_configs of ALL active CLs (of the given ucc_lib), if one fails (param not found in some CL) than the error is returned. 

2. ucc_context_config_modify(cfg, "opt,basic", "NET_DEVICES", "mlx5_0:1") will try to set "NET_DEVICES" param for "opt" and "basic" CLs.

### CODE REVIEW

**Comment by manjugv** [comment]
@vspetrov Is this ready for review?

**Review by alex--m — COMMENTED** [review]
@vspetrov would it be fair to assume the code is modified UCX code? it's a lot of code to review, and I think it's safe to assume UCX's code is ok...

**Comment by vspetrov** [comment]
> @vspetrov would it be fair to assume the code is modified UCX code? it's a lot of code to review, and I think it's safe to assume UCX's code is ok...

@alex--m not really, the code uses UCX indeed but it is not copy paste. Most logic is: each CL registers ctx_config table. Then when ucc_ctx_config_read would loop through the list of selected CLs and call all the corresponding ucs_config_parser_fill_opts (for each cl config). Basically this infrastructure code is what this PR is about. Additionally, plz note the change in the API: ucc.h.

btw, PR is not that large imho

File: src/core/ucc_context.c [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+void ucc_context_config_print(const ucc_context_config_h config, FILE *stream,
+                              const char *title,
+                              ucc_config_print_flags_t print_flags)
+{
+    int i;
+    int print_header = print_flags & UCC_CONFIG_PRINT_HEADER;
+    int flags        = print_flags;
+    /* cl_context_configs will always be printed with HEADER using
+       CL name as title */
+    flags |= UCC_CONFIG_PRINT_HEADER;
```
Comment (manjugv) [inline src/core/ucc_context.c:184]: @vspetrov I'm a bit confused here. Are you trying to print header for each context that belongs to different CL ?

Comment (vspetrov) [reply src/core/ucc_context.c:184]: @manjugv There will a single header line printed as defined by the user (title argument). And then the name of CL printed as header will separate the output - it is more convenient, like sections of parameters for each CL. See example:

```
#
# USER TITLE HEADER
#



#
# CL_BASIC
#

#
# For dbg test purpuse : don't commit
#
# syntax:    unsigned integer
#
UCC_CL_BASIC_TEST_PARAM=5

#
# CL_OPT
#

#
# allreduce knomial radix
#
# syntax:    unsigned integer
#
UCC_CL_OPT_ALLREDUCE_KN_RADIX=4
```

The CL headers in the example above are : "CL_BASIC" and "CL_OPT".
We can remove this and just leave the list of params, they still have the prefix "UCC_<CL_NAME>" so stil can navigate. What do you think guys?

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/cl/basic/cl_basic_lib.c [hunk 2]
```diff
@@ -58,6 +69,13 @@ ucc_cl_basic_iface_t ucc_cl_basic = {
             .table  = ucc_cl_basic_lib_config_table,
             .size   = sizeof(ucc_cl_basic_lib_config_t),
         },
+    .super.cl_context_config =
+        {
+            .name   = "CL_BASIC",
+            .prefix = "CL_BASIC_",
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/cl/basic/cl_basic_lib.c]: added a macro to declare the iface struct

File: src/cl/ucc_cl.h [hunk 3]
```diff
@@ -26,14 +27,21 @@ typedef struct ucc_cl_lib_config {
     int                        priority;
 } ucc_cl_lib_config_t;
 
+typedef struct ucc_cl_context_config {
+    ucc_cl_iface_t *iface;
+    ucc_cl_lib_t   *cl_lib;
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/cl/ucc_cl.h:32]: lib is enough. iface was only for a convenience shortcut (to reduce the chain of "->"). do you think we should remove it?

**Review by manjugv — APPROVED** [review]

**Comment by vspetrov** [comment]
@alex--m do you have any more comments?

**Review by alex--m — APPROVED** [review]
